components/utils.py

```diff
@@ -520,3 +606,282 @@ def untar(source, destination='/tmp', strip=1):
+                    # update the resource file we hold that might have changed
+                    install_resource = self._get_resource_file_path(
+                            service_name, resource_name, False)
+                    copy(existing_resource_path, install_resource)
+                    # copy it to upgrade-resource_name
+                    copy(existing_resource_path, dest)
+                else:
+                    ctx.logger.info('User resource {resource} not found on'
+                                    ' {path}'.format(resource=resource_name,
+                                                     path=dest))
+
+        if not os.path.isfile(dest):
+            self._download_resource(source, dest, params)
+
+    @staticmethod
+    def _download_resource(source, dest, params):
+        resource_name = os.path.basename(dest)
+        ctx.logger.info('Downloading resource {resource} to '
+                        '{dest}'.format(resource=resource_name, dest=dest))
+        ctx.download_resource_and_render(source, dest, params)
+
+    def _get_resource_file_path(self, service_name, resource_name, upgrade):
+        base_service_dir = os.path.join(self.RESOURCES_PATH, service_name)
+        if upgrade:
+            upgrade_resource_name = '{0}{1}'\
+                .format(self.UPGRADE_RESOURCE_PREFIX, resource_name)
+            dest_file_path = os.path.join(base_service_dir,
+                                          upgrade_resource_name)
+        else:
+            dest_file_path = os.path.join(base_service_dir, resource_name)
+        return dest_file_path
+
+    def get_upgrade_resources_json(self, service_name):
+        return self._get_resources_json(service_name, True)
+
+    def get_install_resources_json(self, service_name):
+        return self._get_resources_json(service_name, False)
+
+    def _get_resources_json(self, service_name, upgrade):
+        resources_json = self._get_resource_file_path(service_name,
+                                                      self.RESOURCES_JSON_FILE,
+                                                      upgrade)
+        if os.path.isfile(resources_json):
+            with open(resources_json) as f:
+                return json.load(f)
+        return {}
+
+    def _set_resources_json(self, resources_dict, service_name, upgrade):
+        resources_json = self._get_resource_file_path(service_name,
+                                                      self.RESOURCES_JSON_FILE,
+                                                      upgrade)
+        write_file(resources_dict, resources_json)
+
+    def archive_resources(self, service_name):
+        rollback_dir = os.path.join(self.RESOURCES_ROLLBACK_PATH, service_name)
+        mkdir(rollback_dir)
+
+        service_archive_path = os.path.join(self.RESOURCES_PATH, service_name)
+        # nodes will not always use blueprint resources i.e. java/python
+        if os.path.isdir(service_archive_path) and self.is_upgrade:
```

Comment:
maybe do `if self.is_upgrade` first so that it doesn't check for the dir every time when it doesn't need to?